### PR TITLE
Add cast between json and primitive casts

### PR DESCRIPTION
**Pull request by haozhun**
Depends on ~~#3320~~(merged)

### CODE REVIEW

**Comment by haozhun** [comment]
Dependency merged. Rebased.

**Comment by haozhun** [comment]
@cberner @dain Take a look at the tests first and see if
* you can think of more corner cases
* the output is reasonable.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java [hunk 1]
```diff
@@ -13,37 +13,295 @@
...
+import static com.facebook.presto.spi.type.StandardTypes.VARCHAR;
+import static com.facebook.presto.util.Failures.checkCondition;
+import static com.fasterxml.jackson.core.JsonFactory.Feature.CANONICALIZE_FIELD_NAMES;
+import static java.lang.String.format;
 
 public final class JsonOperators
 {
+    private static final JsonFactory JSON_FACTORY = new JsonFactory().disable(CANONICALIZE_FIELD_NAMES);
+
+    public static void main(String[] args)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonOperators.java]: remove this

**Comment by cberner** [comment]
I think we should survey other databases/libraries to see how they handle JSON-to-primitive conversions. We generally don't do auto-coercions of data in Presto. For example, unlike Hive, `1 != '1'` in Presto. However, with the current implementation of these casts it's hard to have type safety in your JSON structures. For example, suppose I have a logging framework that logs a struct `{userid: 123}`. If I make a mistake and log it as `{userid: "123"}` everything will work fine until one day when I log `{userid: "abc"}`.

Also, I think it makes it harder to rationalize why `CAST(JSON '1' AS VARCHAR)` works, but `CAST(JSON '[1]' AS VARCHAR)` results in an error.

All that said, if this is a common way of handling JSON conversions then I'm fine with it

**Comment by cberner** [comment]
Had one comment. Also, lets not land this until at least version 0.118. Otherwise, looks good
